WordPress/Classes/Services/NotificationSyncMediator.swift

```diff
@@ -312,30 +314,30 @@ private extension NotificationSyncMediator {
     ///     - completion: Callback to be executed on completion
     ///
     func determineUpdatedNotes(with remoteHashes: [RemoteNotification], completion: @escaping (([String]) -> Void)) {
-        let derivedContext = type(of: self).sharedDerivedContext(with: contextManager)
-
-        derivedContext.perform {
-            let remoteIds = remoteHashes.map { $0.notificationId }
-            let predicate = NSPredicate(format: "(notificationId IN %@)", remoteIds)
-            var localHashes = [String: String]()
-
-            for note in derivedContext.allObjects(ofType: Notification.self, matching: predicate) {
-                localHashes[note.notificationId] = note.notificationHash ?? ""
-            }
+        Self.operationQueue.addOperation(AsyncBlockOperation { [contextManager] done in
+            contextManager.performAndSave { context in
+                let remoteIds = remoteHashes.map { $0.notificationId }
+                let predicate = NSPredicate(format: "(notificationId IN %@)", remoteIds)
+                var localHashes = [String: String]()
+
+                for note in context.allObjects(ofType: Notification.self, matching: predicate) {
+                    localHashes[note.notificationId] = note.notificationHash ?? ""
+                }
 
-            let filtered = remoteHashes.filter { remote in
-                let localHash = localHashes[remote.notificationId]
-                return localHash == nil || localHash != remote.notificationHash
-            }
+                let filtered = remoteHashes.filter { remote in
+                    let localHash = localHashes[remote.notificationId]
+                    return localHash == nil || localHash != remote.notificationHash
+                }
 
-            derivedContext.reset()
+                let outdatedIds = filtered.map { $0.notificationId }
 
-            let outdatedIds = filtered.map { $0.notificationId }
+                DispatchQueue.main.async {
+                    completion(outdatedIds)
+                }
 
-            DispatchQueue.main.async {
-                completion(outdatedIds)
+                done()
```

Comment:
Any idea why `completion` would need to be called on the main queue in the original code?

Would it make sense to call `done()` after `completion` in that `async` block? Just so it's really called at the end of the whole `AsyncBlockOperation`?

Reply:
Alternatively, I just noticed in the diff below that we have a similar scenario in which we call `done()` and the input `completion` as part of the `completion` parameter of `performAndSave`. Personally, that looks like the neater option, but I'm still unsure whether `done()` should be called before or after the input `completion` 🤔

Reply:
> Any idea why completion would need to be called on the main queue in the original code?

I have no idea 🥲 . This code is 6 years old, and I couldn't find anything useful using git blame.

> but I'm still unsure whether done() should be called before or after the input completion

I don't think the order would make any difference, and it shouldn't.

When `completion` is called, it should mean the current task is completed, the logic within the `completion` block isn't part of the operation. So, in theory, it shouldn't matter if `done` is called before or after `completion`.

Also, the operation queue is used to prevent concurrent access to the Core Data database via a context object, so it should be safe to mark the current operation as done as soon as we've done using the context object.